### PR TITLE
Feature/terraform for registry

### LINKED ISSUE #5: Create CDK code for sunodo-registry S3
The CDK code should:

- [x] create `sunodo-registry` bucket
- [x] create `sunodo-registry-s3` user
- [x] attach a [restricted Policy](https://github.com/docker/docs/blob/main/registry/storage-drivers/s3.md#s3-permission-scopes) to that user
- [ ] update the sunodo-registry Fly App configuration

### PR DESCRIPTION
**Pull request by oap75**
/close #5 

This PR will deploy S3 bucket for sunodo registry with terraform . 

Resources : 

https://developer.hashicorp.com/terraform/tutorials/aws-get-started/infrastructure-as-code

https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/s3_bucket

### CODE REVIEW

**Comment by tuler** [comment]
Please reformat the code to a consistent style format. My VSCode reformatted the files, to a 2 space indentation, and aligning key/values.

**Comment by tuler** [comment]
Instead of calling `s3` I think we should call the module `registry`, and implement the necessary resources for the registry in that (i.e. a S3 bucket).

**Comment by oap75** [comment]
> Instead of calling `s3` I think we should call the module `registry`, and implement the necessary resources for the registry in that (i.e. a S3 bucket).

I see, Okay I'll fix it then ...

**Comment by endersonmaia** [comment]
> Instead of calling `s3` I think we should call the module `registry`, and implement the necessary resources for the registry in that (i.e. a S3 bucket).

You made me think about the organization of IaC code, should they be together with the module code?

for ex. should we have this terraform code inside the `apps/registry/`?

**Comment by tuler** [comment]
Yes, maybe so. But we have to see how it works with terraform, how it correlates to other pieces of infrastructure, like api

**Comment by oap75** [comment]
@endersonmaia I need to access to the iam policy section to create or remove users. After that I can test my code : 

```
│ Error: creating IAM Policy sunodo_policy: AccessDenied: User: arn:aws:sts::685362664271:assumed-role/AWSReservedSSO_SystemAdministrator_c4699232fd687bad/omid.asadpour is not authorized to perform: iam:CreatePolicy on resource: policy sunodo_policy because no identity-based policy allows the iam:CreatePolicy action
│ 	status code: 403, request id: 1c3cf8df-299f-42bd-a664-2757d802e920
│ 
│   with module.registry.aws_iam_policy.sunodo_iam_policy,
│   on registry/bucket.tf line 38, in resource "aws_iam_policy" "sunodo_iam_policy":
│   38: resource "aws_iam_policy" "sunodo_iam_policy" {

```

**Comment by oap75** [comment]
**Enabled Server Side Encryption:**

Server side encryption protects data at rest. Amazon S3 encrypts each object with a unique key.

Server side encryption encrypts only the object data, not object metadata.

File: terraform/registry/bucket.tf [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+resource "aws_s3_bucket_acl" "sunodo_bucket_acl" {
+  bucket = aws_s3_bucket.sunodo_bucket.id
+  acl    = "private"
+}
+
+resource "aws_s3_bucket_public_access_block" "sunodo_bucket_block_pub_access" {
+    bucket = aws_s3_bucket.sunodo_bucket.id
+    block_public_acls   = true
+    block_public_policy = true
+    restrict_public_buckets = true
```
Comment (endersonmaia) [inline terraform/registry/bucket.tf:20]: Here you use 4 spaces, and 2 elsewhere

Let's stick to a pattern. 

Is there any linter for HCL that we could use within VSCode to make this formating automatic?

File: terraform/registry/bucket.tf [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+ }
+
+resource "aws_s3_bucket" "sunodo_log_bucket" {
```
Comment (endersonmaia) [inline terraform/registry/bucket.tf:23]: Who is gonna use this `sunodo_log_bucket`?

Let's keep the scope of this PR just for the sunodo-registry S3 Bucket

File: terraform/registry/README.md [hunk 3]
```diff
@@ -0,0 +1,78 @@
+# sunodo s3 deployment
```
Comment (endersonmaia) [inline terraform/registry/README.md:1]: Use `markdownlint` here, and fix the problems.

I'll do another review when the linting is done.

**Review by endersonmaia — CHANGES_REQUESTED** [review]
Let's puts this code inside the `./apps/registry/` in a directory named `tf` or `terraform`, or even `infra`.

**Comment by endersonmaia** [comment]
@oap75 is there any code for the diagram?

Let's try to use mermaid whenever possible.

File: terraform/registry/bucket.tf (hunk 1 above)
Comment (oap75) [reply terraform/registry/bucket.tf:20]: Yes, I know there is a TFLint that can lint our TF configs and we even can configure GH action for it ...

Comment (oap75) [reply terraform/registry/bucket.tf:20]: VSCODE extension : https://marketplace.visualstudio.com/items?itemName=HashiCorp.terraform

File: terraform/registry/bucket.tf (hunk 2 above)
Comment (oap75) [reply terraform/registry/bucket.tf]: Fixed !

File: terraform/registry/README.md (hunk 3 above)
Comment (oap75) [reply terraform/registry/README.md]: Fixed !

File: terraform/apps/registry/bucket.tf [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+      "s3:ListBucket",
+      "s3:GetBucketLocation",
+      "s3:ListBucketMultipartUploads",
+    ]
+    resources = [
+      aws_s3_bucket.sunodo_bucket.arn,
+      "${aws_s3_bucket.sunodo_bucket.arn}/*",
+    ]
+  }
+}
```
Comment (endersonmaia) [inline terraform/apps/registry/bucket.tf:41]: This policy is different from the one defined at the suggested link

You see it has two statements, one for the `BUCKET_NAME`, and other for `BUCKET_NAME/*` (its contents).

https://github.com/docker/docs/blob/main/registry/storage-drivers/s3.md#s3-permission-scopes

Please make the proper adjustments.

**Review by endersonmaia — CHANGES_REQUESTED** [review]
I'll send a sub-PR with a siplification, merge it into this PR if you agree.

**Comment by endersonmaia** [comment]
We'll need `AWS_ACCESS_KEY_ID` and `AWS_SECRET_ACCESS_KEY` to configure the registry, and AFAIK we need an IAM User for this. I'm missing this IAM User creation at the tf code.

Also, could that information be an output from `terraform apply`, so that I can define the variables with `flyctl secrets set...`.

File: terraform/apps/registry/bucket.tf (hunk 4 above)
Comment (oap75) [reply terraform/apps/registry/bucket.tf]: Fixed !

File: apps/registry/tf/bucket.tf [hunk 5]
```diff
@@ -0,0 +1,88 @@
+resource "aws_s3_bucket" "sunodo_bucket" {
+  bucket        = var.bucket_name
+  force_destroy = false
+
+  tags = {
+    Name        = "sunodo_bucket"
+    Environment = "Dev"
```
Comment (endersonmaia) [inline apps/registry/tf/bucket.tf:7]:
```suggestion
    Environment = "Production"
```

File: apps/registry/tf/bucket.tf [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+    resources = [
+      "${aws_s3_bucket.sunodo_bucket.arn}/*",
+    ]
+  }
+}
+
+resource "aws_iam_policy" "sunodo_registry" {
+  name        = "sunodo_registry"
+  path        = "/"
+  description = "Sunodo Test Policy"
```
Comment (endersonmaia) [inline apps/registry/tf/bucket.tf:53]:
```suggestion
  description = "Sunodo Registry S3 Storage Policy"
```

File: apps/registry/tf/bucket.tf [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+  policy      = data.aws_iam_policy_document.sunodo_registry.json
+}
+
+resource "aws_iam_role" "sunodo_registry" {
+  name        = "sunodo_registry"
+  description = "test - access to source and destination S3 bucket"
```
Comment (endersonmaia) [inline apps/registry/tf/bucket.tf:59]:
```suggestion
  description = "access to source and destination S3 bucket"
```

File: apps/registry/tf/bucket.tf (hunk 5 above)
Comment (oap75) [reply apps/registry/tf/bucket.tf]: Fixed !

File: apps/registry/tf/bucket.tf [hunk 8]
```diff
@@ -0,0 +1,88 @@
+resource "aws_s3_bucket" "sunodo_bucket" {
+  bucket        = var.bucket_name
+  force_destroy = false
+
+  tags = {
+    Name        = "sunodo_bucket"
```
[the remark this answers is not in the document]
Comment (oap75) [reply apps/registry/tf/bucket.tf]: Fixed !

File: apps/registry/tf/bucket.tf (hunk 6 above)
Comment (oap75) [reply apps/registry/tf/bucket.tf]: Fixed !

File: apps/registry/tf/bucket.tf (hunk 7 above)
Comment (oap75) [reply apps/registry/tf/bucket.tf]: Fixed !

**Review by endersonmaia — CHANGES_REQUESTED** [review]
I see you created the IAM User and its Access Key, but I still don't see any documentation on how to get them and define the secrets for the sunodo-registry Fly App.

Also, the S3 Bucket still shows that Objects can be public, we should make it similar to the other s3 bucket.

image

**Comment by oap75** [comment]
> Also, the S3 Bucket still shows that Objects can be public, we should make it similar to the other s3 bucket.

Fixed !

File: apps/registry/README.md [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+    --org sunodo \
+    set \
+    REGISTRY_STORAGE_S3_ACCESSKEY=... \
+    REGISTRY_STORAGE_S3_SECRETKEY=... \
+```
+
+## Quick Destroy
+
+```shell
+flyctl apps destroy --app sunodo-registry --org sunodo
```
[the remark this answers is not in the document]
Comment (oap75) [reply apps/registry/README.md]: Fixed !

File: apps/registry/README.md [hunk 10]
```diff
@@ -0,0 +1,33 @@
...
+
+## Requirements
+
+- fly.io account;
+- flyctl CLI installed and configured;
+
+## Quick Start
+
+```shell
+flyctl apps deploy --app sunodo-registry --org sunodo
```
[the remark this answers is not in the document]
Comment (oap75) [reply apps/registry/README.md]: Fixed !

File: apps/registry/tf/.gitignore [hunk 11]
```diff
@@ -0,0 +1 @@
+.terraform
```
Comment (endersonmaia) [inline apps/registry/tf/.gitignore:1]: from https://github.com/github/gitignore/blob/main/Terraform.gitignore

```suggestion
# Local .terraform directories
**/.terraform/*

# .tfstate files
*.tfstate
*.tfstate.*

# Crash log files
crash.log
crash.*.log

# Exclude all .tfvars files, which are likely to contain sensitive data, such as
# password, private keys, and other secrets. These should not be part of version 
# control as they are data points which are potentially sensitive and subject 
# to change depending on the environment.
*.tfvars
*.tfvars.json

# Ignore override files as they are usually used to override resources locally and so
# are not checked in
override.tf
override.tf.json
*_override.tf
*_override.tf.json

# Include override files you do wish to add to version control using negated pattern
# !example_override.tf

# Include tfplan files to ignore the plan output of command: terraform plan -out=tfplan
# example: *tfplan*

# Ignore CLI configuration files
.terraformrc
terraform.rc
```

**Review by endersonmaia — CHANGES_REQUESTED** [review]

File: apps/registry/tf/.gitignore (hunk 11 above)
Comment (oap75) [reply apps/registry/tf/.gitignore]: Fixed !

**Comment by oap75** [comment]
> I see you created the IAM User and its Access Key, but I still don't see any documentation on how to get them and define the secrets for the sunodo-registry Fly App.

Fixed !

**Review by endersonmaia — APPROVED** [review]

**Pull request merged**